esp-hal/src/soc/mod.rs

```diff
@@ -85,9 +91,7 @@ pub(crate) fn is_slice_in_dram<T>(slice: &[T]) -> bool {
 pub(crate) fn is_valid_psram_address(address: u32) -> bool {
     #[cfg(psram)]
     {
-        let start = crate::psram::psram_vaddr_start() as u32;
-        let end = start + crate::psram::PSRAM_BYTES as u32;
-        (start..=end).contains(&address)
+        unsafe { (MAPPED_PSRAM_START..=MAPPED_PSRAM_END).contains(&(address as usize)) }
```

Comment:
AtomicRange doesn't exist unfortunately - I believe this has to be a critical section, or we have to accept that it may fail in very specific criminal circumstances.

Reply:
Maybe this shouldn't be a runtime thing anyways - what about just choosing the start and max end address? This check is only needed to know whether something could be in psram, but whatever operation we're about to perform after that is going to fail if psram isn't mapped that far :thinking:. I think that's what idf does?

Reply:
I also thought it's not that risky and a full critical-section is too expensive - but I am thinking about creating something around `esp_hal::LockState` 🤔

Reply:
> Maybe this shouldn't be a runtime thing anyways - what about just choosing the start and max end address? This check is only needed to know whether something could be in psram, but whatever operation we're about to perform after that is going to fail if psram isn't mapped that far 🤔. I think that's what idf does?

Probably this should be `is_valid_extmem_address` - then we should use a chip specific static range.
Does it make any difference for DMA if the external memory is RAM or FLASH?

Reply:
> Does it make any difference for DMA if the external memory is RAM or FLASH?

On the S3 and S2 yes, we need to make sure it doesn't silently fail. Idk if it's practical or possible to differentiate. 

Considering this is initialized once, I think you can get away with using AtomicBool and UnsafeCell. Similar to what Static cell does internally.

Reply:
> Considering this is initialized once

There's nothing currently preventing me from initializing PSRAM multiple times, or is there?

Reply:
After reading the TRM, DMA from FLASH doesn't work and especially 

```rust
pub(crate) fn is_valid_memory_address(address: u32) -> bool {
    is_valid_ram_address(address) || is_valid_psram_address(address)
}
```

Would give us wrong results then and we would silently fail again. Then we can't just use fixed addresses 😞

Reply:
> There's nothing currently preventing me from initializing PSRAM multiple times, or is there?

Oh yeah - should change that